python/dolfinx/fem/function.py

```diff
@@ -312,6 +312,19 @@ def _(u_ptr):
 
         _interpolate(u)
 
+    def interpolate_cells(self, ufl_expression: ufl.core.expr.Expr, cells: np.ndarray = None):
```

Comment:
I'm not sure what this function name means?

Reply:
I still don't understand what `interpolate_cells` means. What is the meaning of `_cells`?

Reply:
That you can interpolate on a subset of cells. If you supply the cell indices (local to process) as a list, you only perform integration over those entities

Reply:
Could be `interpolate_over_cells`, or simply `interpolate`. As i said, I only gave it a distinct name to avoid unclear behavior in the Python layer

Reply:
> I.e.
> 
> ```
> u = Function(V)
> w = Function(W)
> u.interpolate(w)
> u.interpolate_cells(w)
> ```
> 
> would produce the same result with very different code paths.

I think it would be better if you do not put this interpolation into Function's class. If you keep it `Expression.interpolate` it makes things more clear.

Reply:
How would you make that work? Then the user has to supply the quadrature points for the space they are interpolating in to, which I think is rather unnecessary.

Reply:
I would like to have `Expression.interpolate(target_function, cells)`.

The explicitness of creating an Expression is good, because construction of Expression (at least in Python) involves potentially expensive JIT compilation. It is very dangerous to for user since
```
u.interpolate(f)
u.interpolate_cells(f)  # This calls FFCx and JIT!
```

Reply:
My main issue is that you have to pass in the interpolation points to the `Expression`. What happens if a user passes in something else than the interpolation points? The code would then randomly crash at application of the interpolation matrix.

In clear-text, the interface you suggest would look like:
```
expr = Expression(ufl_form, interpolation_points)
expr.interpolate(u, cells)
```
which leaves it up to the user to define the correct interpolation points.

Reply:
Hi Jorgen, this is a good patch, but it's brought up some tangential issues about where everything should be located and what the right interface should be. I'm currently having a think about it.

Reply:
> My main issue is that you have to pass in the interpolation points to the `Expression`. What happens if a user passes in something else than the interpolation points? The code would then randomly crash at application of the interpolation matrix.
> 
> In clear-text, the interface you suggest would look like:
> 
> ```
> expr = Expression(ufl_form, interpolation_points)
> expr.interpolate(u, cells)
> ```
> 
> which leaves it up to the user to define the correct interpolation points.

I do not see the point, your free standing interpolate() has these issues anyway (just needs to add proper sanity checks) . But my main concern is only `Function.interpolate_cells()`, because it is not Function which you interpolate, it is the Function wrapped as an Expression.

We need to encourage users to do the right and fast thing. 
If you want to interpolate dolfinx.Function into another dolfinx.Function just call `u.interpolate(f)` - doesn't involve compilation and is faster.

Reply:
> I do not see the point, your free standing interpolate() has these issues anyway (just needs to add proper sanity checks)

The free-standing interpolation function is not exposed to python, so users won't be able to access it. It can be put in an anonymous namespace if we don't want people to use it.


> We need to encourage users to do the right and fast thing.
> If you want to interpolate dolfinx.Function into another dolfinx.Function just call `u.interpolate(f)` - doesn't involve compilation and is faster.

I am experimenting with how to get single-dispatch to handle this nicely. i.e. default to the non-JIT interpolate if it is only a function that is supplied, but use JIT for anything else, such as linear combinations of functions from different spaces.

Reply:
My two cents:
```
V = FunctionSpace(mesh, ("Lagrange", 2))
f = Function(V)
fem.function.interpolate(lambda x: A * k0**2 * np.cos(k0 * x[0]) * np.cos(k0 * x[1]), f)
```

```
V = FunctionSpace(mesh, ("Lagrange", 2))
f = Function(V)
g = Function(V)
fem.function.interpolate(g, f)
```

```
V = FunctionSpace(mesh, ("Lagrange", 2))
f = Function(V)
# TODO: Add explicit code for interpolation_points
expr = Expression(ufl_form, interpolation_points)
fem.expression.interpolate(expr, f)
```